File: src/nlp/intent_classifier.py
"""
Intent Classifier Module - Strategy Pattern
Provides an abstract base class and concrete implementations
for intent classification using different models.
"""

from abc import ABC, abstractmethod
from typing import Dict, Any, Optional
import torch
from transformers import (
    DistilBertTokenizerFast,
    DistilBertForSequenceClassification,
    pipeline,
)


class IntentClassifier(ABC):
    """
    Abstract base class for intent classifiers (Strategy Pattern).
    Allows swapping different classification models without changing
    the DialogueManager code.
    """

    @abstractmethod
    def classify(self, text: str) -> Dict[str, Any]:
        """
        Classify the intent of a user message.

        Args:
            text: The user's input message.

        Returns:
            Dictionary with keys: 'intent' (str), 'confidence' (float).
        """
        pass


class MockIntentClassifier(IntentClassifier):
    """
    Mock implementation for unit testing and development.
    Returns predefined intents without needing a trained model.
    """

    def __init__(self) -> None:
        self._intents = {
            "balance": ["balance", "money", "account", "how much"],
            "transaction": ["transfer", "send", "payment", "pay"],
            "card": ["card", "credit", "debit", "atm"],
            "loan": ["loan", "mortgage", "borrow"],
            "complaint": ["complaint", "issue", "problem", "unhappy", "wrong"],
            "greeting": ["hello", "hi", "hey", "good morning"],
            "goodbye": ["bye", "goodbye", "see you"],
        }

    def classify(self, text: str) -> Dict[str, Any]:
        text_lower = text.lower()
        best_intent = "unknown"
        best_score = 0.0

        for intent, keywords in self._intents.items():
            score = sum(1 for kw in keywords if kw in text_lower)
            if score > best_score:
                best_score = score
                best_intent = intent

        confidence = min(best_score / 3.0, 1.0)
        return {"intent": best_intent, "confidence": round(confidence, 4)}


class BertIntentClassifier(IntentClassifier):
    """
    Intent classifier using a fine-tuned DistilBERT model.
    Uses GPU (CUDA) if available, falls back to CPU.
    """

    def __init__(self, model_path: str) -> None:
        """
        Args:
            model_path: Path to the saved fine-tuned model directory.
        """
        self._device = torch.device(
            "cuda" if torch.cuda.is_available() else "cpu"
        )
        self._tokenizer = DistilBertTokenizerFast.from_pretrained(model_path)
        self._model = DistilBertForSequenceClassification.from_pretrained(
            model_path
        ).to(self._device)
        self._model.eval()

        # Build id-to-label mapping from model config
        self._id2label = self._model.config.id2label
        self._classifier = pipeline(
            "text-classification",
            model=self._model,
            tokenizer=self._tokenizer,
            device=0 if torch.cuda.is_available() else -1,
        )

    @property
    def device(self) -> torch.device:
        return self._device

    @property
    def num_labels(self) -> int:
        return len(self._id2label)

    def classify(self, text: str) -> Dict[str, Any]:
        result = self._classifier(text)[0]
        # Map label to short intent name
        label = result["label"]
        # Handle both formats: "LABEL_0" style or plain intent names
        if label.startswith("LABEL_"):
            intent_id = int(label.split("_")[1])
            intent = self._id2label.get(intent_id, label)
        else:
            intent = label

        return {"intent": intent, "confidence": round(result["score"], 4)}

File: lib/nlp/intent-classifier.ts
/**
 * Intent Classifier Module - TypeScript Port
 * Strategy Pattern: classifies user messages into banking intents.
 * Uses keyword-based matching with scoring (equivalent to MockIntentClassifier).
 * Can be swapped with @xenova/transformers zero-shot classification for higher accuracy.
 */

export interface IntentResult {
  intent: string;
  confidence: number;
}

// Banking intent keywords mapped to intent names
// Mirrors the Python MockIntentClassifier and banking77 dataset categories
const INTENT_KEYWORDS: Record<string, string[]> = {
  balance: [
    "balance", "money", "account", "how much", "funds", "available",
    "current balance", "savings", "checking", "overdraft", "statement",
  ],
  transaction: [
    "transfer", "send", "payment", "pay", "transaction", "wire",
    "move money", "send money", "direct debit", "standing order", "bacs",
  ],
  card: [
    "card", "credit", "debit", "atm", "pin", "chip", "contactless",
    "lost card", "stolen card", "replacement", "block card", "freeze",
    "activate card", "new card", "virtual card",
  ],
  loan: [
    "loan", "mortgage", "borrow", "repayment", "interest rate",
    "apr", "monthly payment", "borrowing", "lending", "overdraft limit",
  ],
  complaint: [
    "complaint", "issue", "problem", "unhappy", "wrong", "error",
    "disappointed", "frustrated", "angry", "terrible", "awful",
    "disgusting", "unacceptable", "ridiculous", "scam",
  ],
  greeting: [
    "hello", "hi", "hey", "good morning", "good afternoon",
    "good evening", "howdy", "greetings", "what's up",
  ],
  goodbye: [
    "bye", "goodbye", "see you", "take care", "cheers",
    "thanks bye", "that's all", "nothing else",
  ],
  account: [
    "open account", "close account", "account details", "update details",
    "change address", "personal details", "kyc", "verification",
    "identity", "proof of address",
  ],
};

/**
 * Classify the intent of a user message using keyword scoring.
 * Returns the best matching intent with a confidence score.
 */
export function classifyIntent(text: string): IntentResult {
  const textLower = text.toLowerCase().trim();
  let bestIntent = "unknown";
  let bestScore = 0;

  for (const [intent, keywords] of Object.entries(INTENT_KEYWORDS)) {
    let score = 0;
    for (const keyword of keywords) {
      if (textLower.includes(keyword)) {
        // Longer keyword matches get higher weight
        score += keyword.split(" ").length;
      }
    }
    if (score > bestScore) {
      bestScore = score;
      bestIntent = intent;
    }
  }

  // Normalize confidence to 0-1 range
  const confidence = Math.min(bestScore / 3, 1);

  return {
    intent: bestIntent,
    confidence: Math.round(confidence * 10000) / 10000,
  };
}